Mathlib/GroupTheory/DivisibleHull.lean

```diff
@@ -0,0 +1,291 @@
+    rw [show (s'.val * t'.val) • s.val • n = (s.val * s'.val) • t'.val • n by
+      simp_rw [smul_smul]
+      ring_nf]
+    rw [h, h']
+    rw [show (t.val * t'.val) • s.val • m' = (s.val * t.val) • t'.val • m' by
+      simp_rw [smul_smul]
+      ring_nf]
+    rw [show (s.val * s'.val) • t.val • n' = (s.val * t.val) • s'.val • n' by
+      simp_rw [smul_smul]
+      ring_nf]
+    exact zsmul_le_zsmul_iff_right (mul_pos s.prop t.prop)
+
+theorem mk_le_mk {m m' : M} {s s' : Submonoid.pos ℤ} : mk m s ≤ mk m' s' ↔ s' • m ≤ s • m' := by rfl
+
+instance : PartialOrder (DivisibleHull M) where
+  le_refl a := by
+    induction a using ind with | mk m s
+    rw [mk_le_mk]
+  le_trans a b c hab hbc := by
+    induction a using ind with | mk ma sa
+    induction b using ind with | mk mb sb
+    induction c using ind with | mk mc sc
+    rw [mk_le_mk, Submonoid.smul_def] at ⊢ hab hbc
+    rw [← zsmul_le_zsmul_iff_right (show 0 < sb.val from sb.prop)]
+    rw [← zsmul_le_zsmul_iff_right (show 0 < sc.val from sc.prop)] at hab
+    rw [← zsmul_le_zsmul_iff_right (show 0 < sa.val from sa.prop)] at hbc
+    rw [smul_comm _ _ ma, smul_comm _ _ mc]
+    rw [smul_comm _ _ mb] at hab
+    exact hab.trans hbc
+  le_antisymm a b h h' := by
+    induction a using ind with | mk ma sa
+    induction b using ind with | mk mb sb
+    rw [mk_le_mk] at h h'
+    rw [mk_eq_mk_iff_smul_eq_smul]
+    exact le_antisymm h h'
+
+noncomputable
+instance : LinearOrder (DivisibleHull M) where
+  le_total a b := by
+    induction a using ind with | mk ma sa
+    induction b using ind with | mk mb sb
+    exact le_total _ _
+  toDecidableLE := Classical.decRel LE.le
+
+theorem mk_lt_mk {m m' : M} {s s' : Submonoid.pos ℤ} : mk m s < mk m' s' ↔ s' • m < s • m' := by
+  simp_rw [lt_iff_not_ge, mk_le_mk]
+
+noncomputable
+instance : IsOrderedAddMonoid (DivisibleHull M) where
+  add_le_add_left a b h c := by
+    induction a using ind with | mk ma sa
+    induction b using ind with | mk mb sb
+    induction c using ind with | mk mc sc
+    rw [mk_le_mk] at h
+    simp_rw [mk_add_mk, mk_le_mk, smul_add, smul_smul]
+    rw [mul_right_comm sc sa sb, add_le_add_iff_left]
+    rw [mul_right_comm sc sa sc, mul_right_comm sc sb sc]
+    rw [mul_smul _ _ ma, mul_smul _ _ mb]
+    exact zsmul_le_zsmul_right (le_of_lt (sc * sc).prop) h
+
+noncomputable
+instance : PosSMulStrictMono ℚ (DivisibleHull M) where
+  elim a ha b c h := by
+    induction b using ind with | mk mb sb
+    induction c using ind with | mk mc sc
+    simp_rw [mk_lt_mk, Submonoid.smul_def] at h
+    simp_rw [qsmul_mk, mk_lt_mk, Submonoid.smul_def, Submonoid.mul_def, smul_smul]
+    simp_rw [mul_right_comm _ _ a.num, mul_smul _ _ mc, mul_smul _ _ mb]
+    refine (zsmul_lt_zsmul_iff_right ?_).mpr h
+    exact mul_pos (Int.natCast_pos.mpr a.den_pos) (Rat.num_pos.mpr ha)
+
+variable (M) in
+/-- The map `m ↦ m / 1` as an `OrderAddMonoidHom`. -/
+@[simps]
+noncomputable
+def mkOrderAddMonoidHom : M →+o DivisibleHull M where
```

Comment:
Is [`AddOreLocalization.numeratorHom`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/GroupTheory/OreLocalization/Basic.html#AddOreLocalization.numeratorHom) useful here? Also, why is this noncomputable?

Reply:
I'll look into OreLocalization again.

> why is this noncomputable?

Likely leftover from old code. Well spotted